wp-includes/sqlite-ast/class-wp-sqlite-driver.php

```diff
@@ -0,0 +1,1093 @@
+	}
+
+	/**
+	 * Executes a MySQL query in SQLite.
+	 *
+	 * @param string $query The query.
+	 *
+	 * @throws Exception If the query is not supported.
+	 */
+	private function execute_mysql_query( WP_Parser_Node $ast ) {
+		if ( 'query' !== $ast->rule_name ) {
+			throw new Exception( sprintf( 'Expected "query" node, got: "%s"', $ast->rule_name ) );
+		}
+
+		$children = $ast->get_child_nodes();
+		if ( count( $children ) !== 1 ) {
+			throw new Exception( sprintf( 'Expected 1 child, got: %d', count( $children ) ) );
+		}
+
+		$ast = $children[0]->get_child_node();
+		switch ( $ast->rule_name ) {
+			case 'selectStatement':
+				$this->query_type = 'SELECT';
+				$query            = $this->translate( $ast->get_child() );
+				$stmt             = $this->execute_sqlite_query( $query );
+				$this->set_results_from_fetched_data(
+					$stmt->fetchAll( $this->pdo_fetch_mode )
+				);
+				break;
+			case 'insertStatement':
+			case 'updateStatement':
+				$this->execute_update_statement( $ast );
+				break;
+			case 'replaceStatement':
+			case 'deleteStatement':
+				if ( 'insertStatement' === $ast->rule_name ) {
+					$this->query_type = 'INSERT';
+				} elseif ( 'replaceStatement' === $ast->rule_name ) {
+					$this->query_type = 'REPLACE';
+				} elseif ( 'deleteStatement' === $ast->rule_name ) {
+					$this->query_type = 'DELETE';
+				}
+				$query = $this->translate( $ast );
+				$this->execute_sqlite_query( $query );
+				$this->set_result_from_affected_rows();
+				break;
+			case 'createStatement':
+				$this->query_type = 'CREATE';
+				$subtree          = $ast->get_child_node();
+				switch ( $subtree->rule_name ) {
+					case 'createTable':
+						$query = $this->translate( $ast );
+						$this->execute_sqlite_query( $query );
+						$this->set_result_from_affected_rows();
+						break;
+					default:
+						throw new Exception(
+							sprintf(
+								'Unsupported statement type: "%s" > "%s"',
+								$ast->rule_name,
+								$subtree->rule_name
+							)
+						);
+				}
+				break;
+			default:
+				throw new Exception( sprintf( 'Unsupported statement type: "%s"', $ast->rule_name ) );
+		}
+	}
+
+	private function execute_update_statement( WP_Parser_Node $node ): void {
+		// @TODO: Add support for UPDATE with multiple tables and JOINs.
+		//        SQLite supports them in the FROM clause.
+
+		$has_order = $node->has_child_node( 'orderClause' );
+		$has_limit = $node->has_child_node( 'simpleLimitClause' );
+
+		/*
+		 * SQLite doesn't support UPDATE with ORDER BY/LIMIT.
+		 * We need to use a subquery to emulate this behavior.
+		 *
+		 * For instance, the following query:
+		 *   UPDATE t SET c = 1 WHERE c = 2 LIMIT 1;
+		 * Will be rewritten to:
+		 *   UPDATE t SET c = 1 WHERE rowid IN ( SELECT rowid FROM t WHERE c = 2 LIMIT 1 );
+		 */
+		if ( $has_order || $has_limit ) {
+			$subquery = 'SELECT rowid FROM ' . $this->translate_sequence(
+				array(
+					$node->get_descendant_node( 'tableReference' ),
```

Comment:
Is there any danger of sourcing a wrong descendant here? E.g. from a subquery? I wonder what's the right balance between shorthands and verbosity. E.g. would a `get_node( [ "whereClause", "columnsList" ] )` be both useful and safe? Or is there, say, 20 possible paths and listing them all would be tedious?

Reply:
> Is there any danger of sourcing a wrong descendant here?

@adamziel Ah, there is! Great point. It also relates to [the question of breadth-first vs depth-first descendant search](https://github.com/WordPress/sqlite-database-integration/pull/164#discussion_r1854073858). With the current breadth-first implementation, it should pick the correct table reference in this case:

```antlr
updateStatement:
    ({serverVersion >= 80000}? withClause)? UPDATE_SYMBOL LOW_PRIORITY_SYMBOL? IGNORE_SYMBOL? tableReferenceList SET_SYMBOL
        updateList whereClause? orderClause? simpleLimitClause?
;
```

I will write some tests for this.

As for the general problem, it can be more complicated. For instance, if the `tableReferenceList` was optional, then it could be ambiguous.

---

> E.g. would a get_node( [ "whereClause", "columnsList" ] ) be both useful and safe? Or is there, say, 20 possible paths and listing them all would be tedious?

I like this! It's more explicit and safer. In this particular case, it only has a single path `tableReferenceList > tableReference`.

(As a side note, it may seem to be all more complex due to the possibility of multiple table references, but thankfully, that's not possible in MySQL: `However, you cannot use ORDER BY or LIMIT with a multiple-table UPDATE`; [link](https://dev.mysql.com/doc/refman/8.4/en/update.html)).

---

Now, if we make the node-querying API more powerful, does it still belong to the node class, or should we extract it somehow? Reflecting more broadly, it reminds me of DOM query selectors, and these are available on every level, so we may be fine with that. I don't want to over-engineer it at this stage, but taking some inspiration from DOM queries could be interesting. Maybe we could add `$node->query( ... )` and `$node->query_all( ... )` methods:

1. Child: `$node->query( 'tableReferenceList' )`
2. Path: `$node->query( [ 'tableReferenceList', 'tableReference' ] )`
3. Wildcard: `$node->query( ['tableReferenceList', '*', 'tableReference'] )`
4. Recursive search: `$node->query( ['tableReferenceList', '**', 'tableReference'] )`

At the same time, it is a question what would be a reasonable balance between API simplicity and power.